### PR TITLE
Fix for payment method & payment intent

### PR DESCRIPTION
**Pull request by ptibogxiv**

### CODE REVIEW

File: htdocs/install/mysql/tables/llx_societe_rib.sql [hunk 1]
```diff
@@ -23,6 +23,10 @@
 create table llx_societe_rib
 (
   rowid          integer AUTO_INCREMENT PRIMARY KEY,
+
+  --For Stripe or other external payment method
+  ref varchar(128),								-- 'card_...', 'src_...' or 'pm_...'
```
Comment (eldy) [inline htdocs/install/mysql/tables/llx_societe_rib.sql:28]: "ref" is reserved for internal dolibarr ref.
Can you use
"ext_payment_mode_id"
instead ?

Comment (ptibogxiv) [reply htdocs/install/mysql/tables/llx_societe_rib.sql:28]: @eldy  i need it for fetch() in class/companypaymentmode.class.php it use dolibarr fetchcommon class.

So we need to code a new fetch ?

**Comment by ptibogxiv** [comment]
@eldy need to upgrade stripe-php api to  6.34.5 (fix lack of webhook object as payment method)

File: htdocs/install/mysql/tables/llx_societe_rib.sql (hunk 1 above)
Comment (eldy) [reply htdocs/install/mysql/tables/llx_societe_rib.sql:28]: Yes. The common fetch can be used when we need a fetch on internal id or internal ref.

So, if we need a fetch on a field that is not an internal id or ref, it means we must have a specific implementation for the fetch
fetch($id, $ref=null, $ext_payment_mode_id=0)
(The 2 parameters must always be id and ref, even if there is no ref, then we can add specific parameters).

Comment (eldy) [reply htdocs/install/mysql/tables/llx_societe_rib.sql:28]: Can you enhance the fetch and rename "ref" into "ext_payment_mode_id" ?

**Comment by eldy** [comment]
I upgrade stripe api to stripe-php api to 6.35.0

**Comment by ptibogxiv** [comment]
> I upgrade stripe api to stripe-php api to 6.35.0
 
Thanks ! i have code for function in webhook for push  payment and  close invoice but don't have time this week to PR it.

**Comment by eldy** [comment]
Note that the code to execute just after a payment is recorded is already defined into the file paymentok.php (code is different if payment was for a free payment, an invoice, or order...)
So if you code something behind a hook, you must share the same code (for example by moving all the code inside the 
"if ($ispaymentok)"
into a function so you can reuse the function into the webhook...

For failing payment, it is not a problem to have code that diverges, but for payment success, we must have 1 and only one common code (it is used by other payment mode than stripe too).

**Comment by ptibogxiv** [comment]
> Note that the code to execute just after a payment is recorded is already defined into the file paymentok.php (code is different if payment was for a free payment, an invoice, or order...)
> So if you code something behind a hook, you must share the same code (for example by moving all the code inside the
> "if ($ispaymentok)"
> into a function so you can reuse the function into the webhook...
> 
> For failing payment, it is not a problem to have code that diverges, but for payment success, we must have 1 and only one common code (it is used by other payment mode than stripe too).

note for stripe with introduction and 3Dsecure 2 after september 2019, webhook must become default 
With webhook,  we can support more payment method and stripe terminal in TakePOS for physical paiement with terminal when it will be available in europe. automatic confirmation need webhook. we can with manual but it will be probably discontinued by stripe and made bad UI for customer.

A more universal fonction is need for support this evolution (with retrocompatility with old card/source for non EU customer and non EU dolibarr's user after september 2019). Payment intent will become the standard implementation for Stripe in next months.

cf https://stripe.com/docs/payments/payment-intents/quickstart#automatic-confirmation-flow

**Pull request closed without merging**